plugin/pkg/shoot/oidc/openidconnectpreset/admission.go

```diff
@@ -0,0 +1,172 @@
+	readyFunc  admission.ReadyFunc
+}
+
+var (
+	_                             = admissioninitializer.WantsSettingsInformerFactory(&OpenIDConnectPreset{})
+	_ admission.MutationInterface = &OpenIDConnectPreset{}
+
+	readyFuncs = []admission.ReadyFunc{}
+)
+
+// New creates a new OpenIDConnectPreset admission plugin.
+func New() (*OpenIDConnectPreset, error) {
+	return &OpenIDConnectPreset{
+		Handler: admission.NewHandler(admission.Create),
+	}, nil
+}
+
+// AssignReadyFunc assigns the ready function to the admission handler.
+func (o *OpenIDConnectPreset) AssignReadyFunc(f admission.ReadyFunc) {
+	o.readyFunc = f
+	o.SetReadyFunc(f)
+}
+
+// SetSettingsInformerFactory gets Lister from SharedInformerFactory.
+func (o *OpenIDConnectPreset) SetSettingsInformerFactory(f settingsinformer.SharedInformerFactory) {
+	oidc := f.Settings().V1alpha1().OpenIDConnectPresets()
+	o.oidcLister = oidc.Lister()
+
+	readyFuncs = append(readyFuncs, oidc.Informer().HasSynced)
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (o *OpenIDConnectPreset) ValidateInitialization() error {
+	if o.oidcLister == nil {
+		return errors.New("missing oidcpreset lister")
+	}
+	return nil
+}
+
+// Admit tries to determine a OpenIDConnectPreset hosted zone for the Shoot's external domain.
+func (o *OpenIDConnectPreset) Admit(a admission.Attributes, out admission.ObjectInterfaces) error {
+	// Wait until the caches have been synced
+	if o.readyFunc == nil {
+		o.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	if !o.WaitForReady() {
+		return admission.NewForbidden(a, errors.New("not yet ready to handle request"))
+	}
+
+	// Ignore all kinds other than Shoot
+	// Ignore all subresource calls
+	// Ignore all operations other than CREATE
+	if len(a.GetSubresource()) != 0 || a.GetKind().GroupKind() != garden.Kind("Shoot") || a.GetOperation() != admission.Create {
+		return nil
+	}
+	shoot, ok := a.GetObject().(*garden.Shoot)
+	if !ok {
+		return apierrors.NewBadRequest("could not convert resource into Shoot object")
+	}
+
+	// Ignore if the Shoot manifest specifies OIDCConfig.
+	if shoot.Spec.Kubernetes.KubeAPIServer != nil && shoot.Spec.Kubernetes.KubeAPIServer.OIDCConfig != nil {
+		return nil
+	}
+
+	oidcs, err := o.oidcLister.OpenIDConnectPresets(shoot.Namespace).List(labels.Everything())
+	if err != nil {
+		return apierrors.NewInternalError(fmt.Errorf("could not list existing openidconnectpresets: %v", err))
+	}
+
+	preset, err := filterOIDCs(oidcs, shoot)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+	// We have an OpenIDConnectPreset, use it.
+	if preset != nil {
+		applier.ApplyOIDCConfiguration(shoot, preset)
+		return nil
+	}
+
+	return nil
+}
+
+func filterOIDCs(oidcs []*settingsv1alpha1.OpenIDConnectPreset, shoot *garden.Shoot) (*settingsv1alpha1.OpenIDConnectPresetSpec, error) {
```

Comment:
Can this be a utility function used by both the `OpenIDConnectPreset` and `ClusterOpenIDConnectPreset` admission plugins?

Reply:
I've tried but I couldn't come up with something good enough. No generics...

Reply:
Ok